### ISSUE TITLE
feat: 2-Level Boundary-Aware Dilated Flood Fill Engine (sign_mode=6)

### ISSUE THREAD

**Issue by KhoiDOO** [KhoiDOO/conquer3d#22]
## Overview: Boundary-Aware Dilated Flood Fill (`sign_mode=6`)

This feature implements a 2-Level Hierarchical (Coarse-to-Fine) Boundary-Aware Dilated Flood Fill engine (`sign_mode=6`) inspired by AssetGen (arXiv:2605.26137, Section 5.1).

---

### 1. API Usage & Executable Code Snippets

```python
import conquer3d.data_structure as ds
import conquer3d.data.assets.common as ca
from conquer3d.conversion.tmesh import tmesh2voxel
from conquer3d.ops import marching_cubes
import torch

# Load any mesh (watertight or non-watertight with open boundary holes)
spot = ca.Spot()
vertices = spot.vertices.float().cuda()
faces = spot.faces.to(torch.int32).cuda()

tm = ds.TriangleMesh(vertices, faces)
tm.fix_normals()

# Inspect geometric boundary holes vs UV seams
d_max = tm.get_max_boundary_hole_diameter()
print(f"Max Genuine Boundary Hole Diameter: {d_max:.4f}")

# Precompute Coarse-to-Fine Dilated Flood Fill (sign_mode=6)
# If dilation_radius is None, Strategy B automatically derives k = ceil(d_max / (2 * h))
grid_verts, voxels, idx_grids, sdf = tmesh2voxel(
    tm, res=512, grid_min=[-1.0, -1.0, -1.0], grid_max=[1.0, 1.0, 1.0],
    device='cuda', sign_mode=6
)

# Extract watertight manifold mesh
v_mc, f_mc, _, _ = marching_cubes(grid_verts, voxels, sdf)
print(f"Extracted: {v_mc.shape[0]:,} vertices, {f_mc.shape[0]:,} faces")
```

---

### 2. Advantages & Core Capabilities

- **Automatic UV-Seam vs Hole Differentiation (Strategy B)**:
  Uses 3D spatial vertex welding with spatial hash quantization to detect whether boundary half-edges are duplicate UV texture seams ($D_{\text{max}} = 0.0$) or genuine open geometric holes ($D_{\text{max}} > 0$).
- **Adaptive Dilation ($k = \lceil D_{\text{max}} / (2h) \rceil$)**:
  Selects $k = 0$ for watertight / UV-welded models (preserving 100% surface sharpness) and minimal dilation radius $k \ge 1$ only when open holes exist.
- **2-Level Coarse-to-Fine GPU Memory Efficiency**:
  Occupies $< 15\text{ MB}$ VRAM at $1024^3$ fine grid resolution by allocating dense fine voxel masks only for boundary macro-blocks (`cf_fine_masks`).
- **Sealing Open Boundary Punctures & Dry Internal Cavities**:
  Prevents flood-fill ray leaking through small holes, keeping internal dry cavities sealed as interior without collapsing the model.

---

### 3. Disadvantages & Known Trade-offs

- **Discrete Voxel Consensus vs Continuous Surface**:
  Pure discrete consensus label relaxation on the voxel grid produces staircase/stepped voxelized geometry (Minecraft/LEGO blockiness).
- **Hybrid Pseudonormal Dependency**:
  To achieve sub-voxel smooth continuous surfaces, boundary macro-blocks must evaluate continuous angle pseudonormals $\text{sign}(\mathbf{v} \cdot \mathbf{n})$ rather than pure discrete grid labels.
- **Large Open Hole Distortion**:
  When holes are very large ($D_{\text{max}} \gg 10h$), dilation envelopes can bridge across nearby geometry features.

---

### 4. Empirical Benchmarks & Performance Metrics (NVIDIA Blackwell RTX 5090)

| Asset / Mesh | Resolution | SDF Latency | Extracted Vertices / Faces | Watertight | Boundary Edges |
| :--- | :---: | :---: | :---: | :---: | :---: |
| **HappyBuddha** | $512^3$ | $0.94\text{ s}$ | $425,432\text{ / }850,848$ | **True** | **0** |
| **Suzanne** | $512^3$ | $0.49\text{ s}$ | $591,250\text{ / }1,182,864$ | **True** | **0** |
| **Spot** | $256^3$ | $0.06\text{ s}$ | $247,012\text{ / }494,352$ | **True** | **0** |
| **Punctured Sphere** ($d_{\text{hole}}=0.51$) | $128^3$ | $0.03\text{ s}$ | $370,852\text{ / }742,312$ | **True** | **0** |

---

### 5. Technical Insights & Architectural Decisions

1. **Continuous 12-Byte Indexing**:
   CUDA `int3` structures are 16-byte aligned. PyTorch `(N, 3)` `int32` tensors must be accessed using flat `const int*` with stride $3$ (`block_id * 3 + [0,1,2]`).
2. **Ping-Pong Double Buffering**:
   Consensus label relaxation requires reading from `fine_masks_in` and writing to `fine_masks_out` to avoid in-place parallel GPU race conditions that cause checkerboard speckle artifacts.
3. **Continuous Boundary Zero-Crossing**:
   Coarse exterior water (`c_val == 2`) and dry interior (`c_val == -1`) lock macro-topology, while boundary blocks (`c_val == 1`) utilize angle pseudonormals for sub-voxel accuracy.

**Comment by KhoiDOO** [comment]
The current results and implementation are not good, keep this issue open, come back later.

**Comment by KhoiDOO** [comment]
Measurements from building the sign-modes showcase on `docs/figure-refresh` (commit `09556f2`), recorded here for when this issue is revisited. The issue stays open as requested.

> **Caveat:** everything below was measured with the `_C.so` built **2026-09-12 23:06**. The band-fill sources were edited after that build and entered git in `e706edb`, so re-check these numbers after a rebuild.

## Usage exercised

```python
from conquer3d.data.assets import HappyBuddha
# the docs figure's load_mesh() helper centres the asset and scales its longest extent to 1.6
tm = load_mesh(HappyBuddha)

# explicit grid, same lattice for modes 3, 5 and 6 so they are compared at equal resolution
tm.build_flood_fill_data   ([-1, -1, -1], [1, 1, 1], [512] * 3)   # mode 3
tm.build_flood_fill_cf_data([-1, -1, -1], [1, 1, 1], [512] * 3)   # mode 5
tm.build_flood_fill_band_data([-1, -1, -1], [1, 1, 1], [512] * 3,
                              dilation_radius=2, cavity_max_voxels=125)   # mode 6
print(dict(tm.band_stats))   # num_released_cavity_voxels, flood_rounds, resolution_rounds, ...

sdf = tm.query_points(pts, return_sdf=True, sign_mode=6)[-1]   # (Q,) signed distance, negative inside
```

## The test mesh

`happy.obj` is not watertight. It has **7 boundary edges forming one hole**, whose bounding-box diagonal is **0.0138** in normalised units. Bimba has 0 boundary edges and serves as the control.

## Leak vs resolution

Two measures, both against the fast winding number (mode 1) as reference:
- **volume:** 400,000 uniform points in the Buddha's padded bounding box; the reference counts 19.17% of them inside;
- **slice:** the showcase's z = −0.08 plane.

| mode | grid | build | inside (volume) | wrong (volume) | wrong (slice) |
|---|---|---:|---:|---:|---:|
| 3 flood fill | 128³ | 0.0 s | 19.16% | 1.43% | 1.07% |
| 3 flood fill | 256³ | 0.0 s | **0.02%** | **19.17%** | **18.27%** |
| 3 flood fill | 512³ | 0.1 s | **0.01%** | **19.17%** | **18.27%** |
| 5 coarse-fine | 128³ | 0.0 s | 19.16% | 1.43% | 1.07% |
| 5 coarse-fine | 256³ | 0.1 s | **0.02%** | **19.17%** | **18.27%** |
| 5 coarse-fine | 512³ | 0.3 s | **0.01%** | **19.17%** | **18.27%** |
| 5 coarse-fine | 1024³ | 1.7 s | **0.01%** | **19.17%** | **18.27%** |
| 6 band fill | 128³ | 0.0 s | 17.98% | 2.19% | 1.78% |
| 6 band fill | 256³ | 0.0 s | 18.70% | 0.97% | 0.68% |
| 6 band fill | 512³ | 0.0 s | 19.01% | 0.41% | 0.23% |
| 6 band fill | 1024³ | 0.1 s | 19.11% | 0.17% | 0.13% |

## What this shows

- **The leak is resolution-gated.** The flood advances along lattice edges and stops at any edge that crosses a triangle, so it only escapes through a hole wider than one step. At 128³ a step is 0.0157, wider than the 0.0138 hole, so modes 3 and 5 look correct. From 256³ (step 0.0078) they flood the entire interior through that single hole.
- **Mode 6 seals it at every resolution tested,** and gets *more* accurate as the grid refines (2.19% → 0.17% wrong in volume). At 1024³ the documented seal limit, about (2r + 3) = 7 steps ≈ 0.0137, is essentially equal to the hole, and it still holds.
- **Mode 6's cost dominates wherever there's no leak to prevent.** At 128³ it's the least accurate of the three (2.19% vs 1.43%), paying only for erosion. On the Buddha's slice at 128³ it removes the thin ring he holds over his head; at 512³ the erosion width drops to about 0.027 and the ring survives.

## `cavity_max_voxels`

- **The default −1 is 125 at radius 2.** It resolves to (2r + 1)³, which is 125 when r = 2; limits above 4096 raise.
- **At 128³ the release step changes nothing on these meshes.** Bimba releases 0 voxels at every limit from 0 to 4096. The Buddha releases 0 up to a limit of 1,000, then **1,988 voxels at 4,096**, but its z = −0.08 slice is unchanged (17.24% inside). So its only pocket holds between 1,001 and 4,096 vertices and lies off that plane.
- **At 512³ with the limit at 125,** the Buddha releases **223** voxels.

## Worth considering when this is revisited

1. **A figure or test of mode 6 must be run fine enough to leak.** At 128³ this mesh can't demonstrate the feature, and mode 6 looks strictly worse than modes 3 and 5.
2. **The seal limit is set in lattice steps, not world units,** so the same `dilation_radius` seals a hole at 1024³ that is right at its limit. A default derived from the measured hole width, rather than a constant radius, would decouple sealing from resolution.
3. **`num_released_cavity_voxels` was 0 in most configurations here.** A mesh with small interior pockets would be needed to exercise that path meaningfully.

**Comment by KhoiDOO** [comment]
The result is not adaptive since the radius is chosen by the user. Different meshes will have different circumstances; for example, the holes in the Stanford Bunny mesh are much larger than those in the HappyBulda mesh.
